### PR TITLE
Protocol implementation for Playwright

### PR DESCRIPTION
**Pull request by sam-ulrich1**
Thanks for contributing to StormCrawler, your efforts are appreciated!

Developer Certificate of Origin
===============================

By contributing to StormCrawler, you accept and agree to the following terms and conditions (the *Developer Certificate of Origin*) for your present and future contributions submitted to StormCrawler. 
Please refer to the *Developer Certificate of Origin* section in [`CONTRIBUTING.md`](CONTRIBUTING.md) for details.

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

Before opening a PR, please check that: 

* You've applied the formatting used by the project with `mvn java-formatter:format`
* You've squashed your commits into a single one
* You've described what the PR does or at least point to a related issue
* You've signed-ff your commits with 'git commit -s'

Thanks!

### CODE REVIEW

**Comment by sam-ulrich1** [comment]
This is a first pass of Playwright implementation for SC.

Points of interest to review:

- How should timeouts be handled? Currently throw an exception
- Stack trace from inner BrowserThread exceptions is not accurate as it is thrown in the calling thread of PlaywrightProtocol.getProtocolOutput. Thoughts?
- DelegationProtocol has a very specific JSON schema and design. We may want to look into if this is the best design for the en user

**Comment by sam-ulrich1** [comment]
@jnioche My last few commits are signed since I set up GitKraken to auto-sign for me. The prior ones are not. I'm not sure how to handle this because I'm having trouble with the tree when I try to backward sign. My apologies, those non-signed commits of the past are coming back to haunt me

**Comment by jnioche** [comment]
That's fab thanks @sam-ulrich1 

I would have separated the Playwright  part from the protocol delegation as they are conceptually quite different. The commits contain changes that are already in the main branch + if possible could you please squash the commits into one?

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/delegate/DelegatorProtocol.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    public void configure(Config conf) {
+        // load delegation filters from the passed configuration
+        protocolFilters = DelegationFilters.fromConf(conf);
+    }
+
+    @Override
+    public ProtocolResponse getProtocolOutput(String url, Metadata metadata)
+            throws Exception {
+        // route robots.txt urls to the robots protocol explicitly
+        if (url.endsWith("/robots.txt"))
```
Comment (jnioche) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/delegate/DelegatorProtocol.java:43]: not sure it is worth making a special case for robots.txt
they can just be processed with the last protocol listed (assuming they don't match with anything further up)

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/okhttp/HttpProtocol.java [hunk 2]
```diff
@@ -263,7 +264,7 @@ public ProtocolResponse getProtocolOutput(String url, final Metadata metadata) t
             SCProxy prox = proxyManager.getProxy(metadata);
 
             // conditionally configure proxy authentication
-            if (StringUtils.isNotBlank(prox.getUsername())) {
+            if (StringUtils.isNotBlank(prox.getAddress())) {
```
Comment (jnioche) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/okhttp/HttpProtocol.java]: Changes on this class are already in the main branch

File: core/src/main/java/com/digitalpebble/stormcrawler/protocol/delegate/DelegationFilters.java [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    protected Protocol defaultProtocol;
+    protected ArrayList<JsonNode> protocolFilters;
+    protected ArrayList<Protocol> protocolImplementations;
+
+    private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(DelegationFilters.class);
+
+    /**
+     * Loads and configure the DelegationFilters based on the storm config.
+     **/
+    public static DelegationFilters fromConf(Config stormConf) {
```
Comment (jnioche) [inline core/src/main/java/com/digitalpebble/stormcrawler/protocol/delegate/DelegationFilters.java:54]: not sure we need a separate class for that - the logic of loading the sub-protocols could be handled in DelegatorProtocol

**Review by jnioche — CHANGES_REQUESTED** [review]
I may have messed things up a bit by deleting the files containing stuff already in master. Let's break things down into 2 separate PRs based on the latest master, it will make it easier to review them. Thanks!

**Comment by sam-ulrich1** [comment]
I can move the Playwright stuff into a separate branch. Is it worth waiting for the DelegatorProtocol merge? If so I can branch from the master there and then confirm that the implementation works both via test and with a full run

**Comment by jnioche** [comment]
> I can move the Playwright stuff into a separate branch. Is it worth waiting for the DelegatorProtocol merge? 

I have merged it, it does not impact any existing code so pretty safe

> If so I can branch from the master there and then confirm that the implementation works both via test and with a full run
yes please start a brand new branch based on the latest master

that will be a great way of testing #900
also could you please modify the test I created there so that it uses your playwright  implementation for the case where  `js: "true"`?

thanks a lot

**Comment by sam-ulrich1** [comment]
No problem. I'll try to get it knocked out by EOD

**Pull request closed without merging**